src/configurationProvider.ts

```diff
@@ -203,8 +206,86 @@ export class JavaDebugConfigurationProvider implements vscode.DebugConfiguration
         return !Array.isArray(configItems) || !configItems.length;
     }
 
-    private async chooseMainClass(folder: vscode.WorkspaceFolder | undefined): Promise<IMainClassOption | undefined> {
-        const res = await resolveMainClass(folder ? folder.uri : undefined);
+    private async resolveLaunchConfig(folder: vscode.Uri | undefined, config: vscode.DebugConfiguration): Promise<IMainClassOption> {
+        if (!config.mainClass) {
+            return await this.promptMainClass(folder);
+        }
+
+        const containsExternalClasspaths = !this.isEmptyArray(config.classPaths) || !this.isEmptyArray(config.modulePaths);
+        const validationResponse = await validateLaunchConfig(folder, config.mainClass, config.projectName, containsExternalClasspaths);
+        if (!validationResponse.mainClass.isValid || !validationResponse.projectName.isValid) {
+            return await this.fixMainClass(folder, config, validationResponse);
+        }
+
+        return {
+            mainClass: config.mainClass,
+            projectName: config.projectName,
+        };
+    }
+
+    private async fixMainClass(folder: vscode.Uri | undefined, config: vscode.DebugConfiguration, validationResponse: ILaunchValidationResponse):
+        Promise<IMainClassOption | undefined> {
+        const errors: string[] = [];
+        if (!validationResponse.mainClass.isValid) {
+            errors.push(String(validationResponse.mainClass.message));
+        }
+
+        if (!validationResponse.projectName.isValid) {
+            errors.push(String(validationResponse.projectName.message));
+        }
+
+        if (validationResponse.proposals && validationResponse.proposals.length) {
+            const answer = await utility.showErrorMessageWithTroubleshooting({
```

Comment:
dup with line 257, you can keep the promise of utility.showErrorMessageWithTroubleshooting and await it if validationResponse.proposals have values.

Reply:
There is a tiny difference between if/else conditions here. For the case of having proposals, the UI shows a `Fix` button, and the debug session will wait the user to click it. For the case of no having proposals, the UI just shows a `Learn More` button, no need to wait the user's click behavior, and the debug session will terminate directly.